### DISCUSSION TITLE
Issues when invoking go functions locally

### DISCUSSION THREAD

**Discussion by paulbkim-dev** [vercel/vercel#10941] [Help]
### Summary

I have serverless functions set up for vercel deployment.
I have no go.work set up.Just go.mod, and no problem working when deployed.
However if I run on vercel dev I get below message as I am invoking the function.

```
go: module . listed in go.work file requires go >= 1.21, but go.work lists go 1.18; to update it:
        go work use
Error: Command failed: go build -ldflags -s -w -o ./vercel-dev-server-go ./...
```

go mod inside the .vercel/cache folder

```
module handler

go 1.21

require (
	github.com/google/uuid v1.4.0
	github.com/jackc/pgx/v5 v5.5.1
	github.com/rs/zerolog v1.31.0
	gopkg.in/mail.v2 v2.3.1
)

require (
	github.com/mattn/go-colorable v0.1.13 // indirect
	github.com/mattn/go-isatty v0.0.20 // indirect
	golang.org/x/sys v0.15.0 // indirect
	gopkg.in/alexcesaro/quotedprintable.v3 v3.0.0-20150716171945-2caba252f4dc // indirect
)

require (
	github.com/jackc/pgpassfile v1.0.0 // indirect
	github.com/jackc/pgservicefile v0.0.0-20221227161230-091c0ba34f0a // indirect
	golang.org/x/crypto v0.16.0 // indirect
	golang.org/x/sync v0.5.0 // indirect
	golang.org/x/text v0.14.0 // indirect
)
require github.com/paulbkim01/paulbkimdev/api v0.0.0-unpublished
replace github.com/paulbkim01/paulbkimdev/api => ../../../../api/
```

### Steps to Reproduce

- get working version of go handlers in api with go.mod version 1.21 
- run locally with astro project `vercel dev`
- invoke the api endpoint to run the handler.

**Comment by wintermute-cell** [comment]
bump, same issue here

**Comment by metomimic** [comment]
has this been resolved yet?

**Comment by sailei1** [comment]
how to fix？

**Reply by sailei1** [reply]
change go.mod
go 1.18.1 can run
But the higher version doesn’t work

**Comment by michael-freling** [comment]
I have the same issue, and I tried to make this works.
Let me share what I've found out

## Possible cause
It seems the error message shows because there is no go version set in `go.work` file, according to [this go issue](https://github.com/golang/go/issues/66207), and `go.work` file is generated by `vercel dev`.
I also confirmed `go.work` file in `.vercel/cache/` file doesn't include go.work, and running `go build` there also has the same error.

```bash
> cat .vercel/cache/go/r80oal2j1qg/go.work
use (
  .
)
> cd .vercel/cache/go/r80oal2j1qg/
> go build -o ./vencel-dev-server-go ./...
go: module . listed in go.work file requires go >= 1.21, but go.work lists go 1.18; to update it:
        go work use
```

And when I added the go version in go.work, it worked

```bash
> grep 'go 1.2' go.mod
go 1.21
> sed -i'' -e '1 i go 1.21' go.work
> cat go.work
go 1.21
use (
  .
)
> go build -o ./vencel-dev-server-go ./...
> ls
api  go.mod  go.work  go.work.sum  vencel-dev-server-go

> ./vencel-dev-server-go
2024/04/28 17:08:56 /home/michael/src/nooxac/go/nooxac-gateway/pkg/db/client.go:74
[error] failed to initialize database, got error dial tcp 127.0.0.1:3306: connect: connection refused
```

## Possible fix

I couldn't find a workaround, but adding the go.version in `go.work` [here](https://github.com/vercel/vercel/blob/26bc51e0d641ea45d304b601848b8123feeade48/packages/go/src/index.ts#L785) should fix the issue.
I also tried to fix only locally like [this](https://github.com/vercel/vercel/commit/68741d49acb06eafb886fda38f79727d2988415d)

Hope it helps the owner of this repository to look into an error and fix an issue faster.

---

By the way, I also see there are a few questions that are similar to this such as followings, so I'm sorry if this isn't the good place.
- https://github.com/orgs/vercel/discussions/5858
- https://github.com/orgs/vercel/discussions/5166
Also, I'm not familiar with this codebase or this project so I might be wrong
